File: programers_Lv1/삼총사.py
# https://school.programmers.co.kr/learn/courses/30/lessons/131705?language=python3
import itertools

# itertools의 순열, 조합 활용
# 이 문제에서는 조합을 사용
# perm = itertools.permutations(chars, 2)  # 순열
# comb = itertools.combinations(chars, 2)  # 조합


def solution(number):
    answer = 0
    result = list(itertools.combinations(number, 3))  # 조합
    for i in result:
        if sum(i) == 0:
            answer += 1
    return answer


print(solution([-2, 3, 0, 2, -5]))  # 2

File: programers_Lv1/삼총사.js
// https://school.programmers.co.kr/learn/courses/30/lessons/131705?language=javascript
function solution(number) {
  var answer = 0;

  //   순열 사용자 정의 함수
  function combinations(arr, k) {
    const result = [];

    function backtrack(subset, start) {
      if (subset.length === k) {
        result.push([...subset]);
        return;
      }

      for (let i = start; i < arr.length; i++) {
        subset.push(arr[i]);
        backtrack(subset, i + 1);
        subset.pop();
      }
    }

    backtrack([], 0);
    return result;
  }

  const result = combinations(number, 3);

  result.forEach((el) => {
    answer += el.reduce((acc, curr) => acc + curr, 0) == 0 ? 1 : 0;
  });

  return answer;
}

console.log(solution([-2, 3, 0, 2, -5]));
